business/authentication/auth_controller.go

```diff
@@ -0,0 +1,47 @@
+package authentication
+
+import (
+	"net/http"
+
+	"github.com/kiali/kiali/config"
+)
+
+// AuthController is the interface that all Kiali authentication strategies should implement.
+// An authentication controller is initialized during Kiali startup.
+type AuthController interface {
+	// Authenticate handles an HTTP request that contains credentials. The method to pass the credentials
+	// is chosen by the authentication controller implementation. The credentials are verified and if
+	// it is supported by the controller, RBAC permissions are verified to ensure that the logging in user
+	// has enough privileges to login to Kiali.
+	// An AuthenticationFailureError is returned if the authentication request is rejected (unauthorized). Any
+	// other kind of error means that something unexpected happened.
+	Authenticate(r *http.Request, w http.ResponseWriter) (*UserSessionData, error)
```

Comment:
Yes. Why you ask? :-)

Reply:
I wasn't sure if `Authenticate` should call `w.WriteHeader(401)` or if it's just expected to return `AuthFailureErr` and the caller handles `w.WriteHeader(401)`.

Reply:
I wanted to do that way and handle the response here. But I missed those `RespondWithJSONIndent`, `RespondWithError` and `RespondWithDetailedError` functions. All these functions are in the `handlers` package, so I can't use them because it would cause an import loop handlers -> authentication -> handlers.

I wanted to extract these funcs to the utils/http folder, but they are widely used in the handlers folder and moving them would cause a lot of churn.

Any better suggestions?

Reply:
This would probably cause a lot of churn too but perhaps moving the `authentication` package to `handlers`? It seems more related to handling auth for http requests and less related to business specific logic anyways.

Reply:
Not a blocker. You can decide if you want to keep it the way it is or if you'd like to change it.

Reply:
I don't think it will cause more churn, because it's a new package.
Anyway, I would like to leave it as is in this PR.

Soon, I'll have ready the OpenID refactor. Once you review the upcoming one, if you still think it belongs to `handlers`, then I can move all this code back to `handlers`.